### ISSUE TITLE
Conditions inside included tasks should not be processed

### ISSUE THREAD

**Issue by jphalip** [ansible/ansible#4251]
This is related to #4074. Consider the following example.

main.yml:

```
- hosts: all

  vars:
    myvar: false

  tasks:
    - include: included.yml
      when: myvar == True
```

included.yml:

```
# This will be skipped.
- command: ls ~

# This will be skipped, too. Therefore the variable 'foo' will not be created.
- command: echo 'hello'
  register: foo

# This will fail since the variable 'foo' hasn't been created in the previous task.
- command: echo 'hi'
  when: foo.rc == 0
```

The last task fails because its `when` clause is processed. Sure, in this particular case one could add a `foo is defined` condition, but my point is that this `when` clause should simply never get processed anyway. The surrounding condition (`when: myvar == True`) should take full precedence and completely disregard any other conditions inside the included tasks.

In other words, the surrounding condition should:
- be added to included tasks that do not already have a `when` clause — this is already happening.
- replace the `when` clause of included that already have their own — this is what I suggest becomes the new behavior.

This would guarantee that all included tasks get skipped when the surrounding condition evaluates to False.

I hope I have explained this clearly enough. Let me know if you'd like me to clarify anything. Thanks!

**Comment by mpdehaan** [comment]
What version of Ansible are you using?

I recall this issue being fixed in 1.3.

**Comment by mpdehaan** [comment]
what happens (should happen) is the first when gets inserted and evaluated prior to the second when, and short circuit evaluation results in the second when statement.

I would (strongly) disagree that it should replace the second as that breaks desired functionality of being able to have a when statement on the include and another on the task.

Anyway, please let me know what "ansible --version" says and if you are using 1.2, testing on 1.3 would be very welcome.

Thanks!

**Comment by jphalip** [comment]
I'm currently using the latest, 1.3.2.

**Comment by jphalip** [comment]
In my mind the `when` clause in the `include` task should take precedence. What is the purpose of evaluating the `when` clauses inside the included tasks if the point is to skip those tasks anyways?

Am I missing something?

**Comment by mpdehaan** [comment]
It's always important to define what "fails" means in a bug ticket so we can get to the point shorter.  Thanks!

When you said it failed can you please show how it failed?

The when clause in the include _should_ take precedence, in fact, that's what I thought it did.

If you mean undefined variable detection still triggers that's another story and possibly unrelated, though I understand what you mean about the basic cause.

Let me know.  Thanks!

**Comment by jphalip** [comment]
Sorry, by 'fails' I mean you see an error like this:

```
fatal: [111.111.111.111] => error while evaluating conditional: {% if foo.rc == 0 %} True {% else %} False {% endif %}
```

Does this help?

**Comment by jphalip** [comment]
I've re-read my original post and I think it was a bit inaccurate and confusing. Please allow me to rephrase more succinctly.

Basically the only thing I'm suggesting is that, when the `include` task's `when` condition evaluates to `False`, any potential `when` conditions inside the included tasks completely get ignored. The reasoning is that one would expect all included tasks to be skipped anyways.

What do you think?

**Comment by mpdehaan** [comment]
The when statement doesn't get evaluated at that time but is affixed to each task therein.

The reason being some values may be of inventory scope.
